design/design-doc-composition.md

```diff
@@ -0,0 +1,816 @@
+problem of multi-cloud; when a claim may provide configuration inputs that may
+be used to match or provision many kinds of managed resource it may support only
+the settings that apply to _all_ compatible managed resources. This is in part
+why Crossplane defines relatively few resource claim kinds. Meanwhile, it's
+possible that an infrastructure operator deploys Crossplane to provide an
+opinionated abstraction for the application operators in their organisation and
+that said organisation only uses AWS. If this organisation values Crossplane's
+separation of concerns but does not need its portability there is no reason that
+its application operators should be limited to resource claims that are
+constrained by supporting all possible providers.
+
+## Goals
+
+## Proposal
+
+This document proposes the introduction of four new Crossplane resource kinds:
+
+* `ApplicationDefinition` - Defines a new kind of Kubernetes custom resource
+  that represents an application.
+* `InfrastructureDefinition` - Defines a new kind of Kubernetes custom resource
+  that represents a logical group of infrastructure.
+* `InfrastructurePublication` - Defines a new kind of Kubernetes custom resource
+  that binds a logical group of infrastructure to an application by "publishing"
+  a resource defined by an `InfrastructureDefinition`.
+* `Composition` - Configures how one or more custom resources should be rendered
+  in response to the creation or modification of a custom resource defined by an
+  `ApplicationDefinition` or `InfrastructureDefinition`.
+
+### Composition
+
+A `Composition` configures how one or more custom resource should be rendered in
+response to the creation or modification of a custom resource defined by an
+`ApplicationDefinition` or `InfrastructureDefinition`. No controller watches for
+the `Composition` kind; a `Composition` is loosely analogous to the existing
+Crossplane concept of a resource class. Unlike resource classes, compositions:
+
+* May define how to compose applications as well as infrastructure.
+* May compose more than one resource, including other composed resources.
+* Explicitly configure their relationship with a defined application or
+  infrastructure resource kind.
+
+Here's an example composition:
+
+```yaml
+apiVersion: crossplane.io/v1alpha1
+kind: Composition
+metadata:
+  name: private-mysql-server
+  labels:
+    connectivity: private
+spec:
+  # This composition declares that its input values will be read 'from' a
+  # resource of the specified kind. The field name 'from' denotes the
+  # relationship with the 'fromFieldPath' notation below.
+  from:
+    apiVersion: database.example.org/v1alpha1
+    kind: MySQLInstance
+  # This composition declares that its input values will be written 'to' the
+  # below resources. The field name 'to' denotes the relationship with the
+  # 'toFieldPath' notation below.
+  to:
+  - base:
+      apiVersion: azure.crossplane.io/v1alpha3
```

Comment:
> No native validation of the resources.

This seems like a tradeoff between quality of validation and number of defined resource kinds. Either way we can validate this YAML, but I agree that it would likely be faster and safer to allow the API server to do so natively. This native validation would come at the cost of generating a CRD for every combination of resources that could be composed to satisfy a defined composite.

Let's take the contemporary `KubernetesCluster` resource claim concept, and assume we'd want to allow that claim to be satisfied by a composition of Kubernetes-ey resources in each of GCP, AWS, and Azure. In this design the infrastructure operator would define `KubernetesCluster` by authoring an `InfrastructureDefinition`. They'd then publish it for use by application operators by authoring an `InfrastructurePublication`, and finally author three `Compositions`; one for each of GCP, AWS, and Azure.

If an application operator was curious which compositions they could use with their `KubernetesClusterBinding` they could run:

```bash
# Determine which compositions can be used with a KubernetesClusterBinding.
# This requires the reader to mentally translate KubernetesClusterBinding to KubernetesCluster.
$ kubectl get compositions
NAME                COMPOSITE-KIND
big-gcp-cluster     KubernetesCluster
big-aws-cluster     KubernetesCluster
big-azure-cluster   KubernetesCluster
small-gcp-database  MySQLInstance
cool-blog           Wordpress
```

In total **five resource kinds were used to implement these compositions**; `InfrastructureDefinition`, `InfrastructurePublication`, `Composite`, `KubernetesCluster`, and `KubernetesClusterBinding`.

Walking through the same use case with [`ClaimDefinition` and `ResourceDefinition`](https://github.com/muvaf/define-your-claim), the infrastructure operator would define `KubernetesCluster` (at both the cluster and namespace scopes) by authoring a `ClaimDefinition`. They would then define three `ResourceDefinitions`; `biggcpclusters.example.org`, `bigawsclusters.example.org`, and `bigazureclusters.example.org`. Finally, they would instantiate one of each of the generated classes by authoring a `BigGCPClusterClass`, `BigAWSClusterClass`, and a `BigAzureClusterClass`.

If an application operator was curious which compositions they could use with their `KubernetesClusterBinding` they could run:

```bash
# First, determine which kinds of resource classes can be used with KubernetesCluster by determining
# which resource definitions exist. This requires the user to mentally translate KubernetesCluster
# to kubernetesclusters.example.org.
$ kubectl get resourcedefinitions
NAME                          CLAIM-NAME
biggcpclusters.example.org    kubernetesclusters.example.org
bigawsclusters.example.org    kubernetesclusters.example.org
bigazureclusters.example.org  kubernetesclusters.example.org
smallgcpdatabases.example.org mysqlinstances.example.org
coolblogs.example.org         wordpresses.example.org

# Second, determine which instances of the desired resource class exists. This requires the user to
# mentally translate biggcpclusters.example.org to BigGCPClusterClass.
$ kubectl get biggcpclusterclasses
NAME
reallybig
kindabig
```

In total **ten resource kinds were used to implement these compositions**; `ClaimDefinition`, `ResourceDefinition`, `KubernetesCluster` (namespaced), `KubernetesCluster` (cluster scoped), `BigGCPCluster`, `BigAWSCluster`, `BigAzureCluster`, `BigGCPClusterClass`, `BigAWSClusterClass`, and `BigAzureClusterClass`.

I agree that introducing these additional concepts makes the latter design safer. Each resource kind can be natively validated by the API server. The separation of functionality into several distinct resource kinds enables granular RBAC roles, and reduces the blast radius of updating some resources. However, having written this all out, I am not convinced that the increased safety of this model justifies the increased cognitive complexity.

Reply:
> This composition targets West US location. If I wanted to have the same class but in a different location, I'll have to duplicate the patches.

I don't think the term "wrong abstraction" in the [articles that describe this concept](https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction), but this is a case where I feel a little duplication is preferable to additional abstraction.

> Then later on, if MySQLInstance schema changes or a new field is added for finer granularity, I have to and fix every instance of Composition object.

If the `MySQLInstance` definition adds a new _required_ field it's making a breaking change, which is going to be painful no matter what; you'd need to update all extant `MySQLInstance` objects too. If `MySQLInstance` is adding a new optional field it doesn't seem so bad to gradually _write new compositions_ that leverage that new field, if they want to.

Note that I mean add new compositions - not edit the existing ones. A composite always uses some set of inputs to produce a set of outputs. The inputs might be a `MySQLInstance`, a `ResourceDefinition`, and a `BigDatabaseClass`, or they might be a `MySQLInstance` and a `Composition`. If we presume those inputs are used to constantly reconcile the `MySQLInstance` with its outputs I think we should seriously consider making all of the other inputs immutable. The `MySQLInstance` author may edit it, and their edits will be reflected on the output resources, but no one may edit the `Composition` or edit _which_ composition the `MySQLInstance` uses. Doing so could lead to large, cascading changes to many instances whether you're editing a `Composition`, editing a `ResourceDefinition`, or editing a `BigDatabaseClass`.

> The alternative is as we talked, separating the actual parameters from the bindings.

You're suggesting decoupling the "base" resource template from the patches that modify them, right? For example the base resources are in a `BigDatabaseClass` and the patches are in a `ResourceDefinition`? This is probably subjective, but I feel that the authoring experience is less awkward and error prone when the patches are specified immediately below the thing they're patching.

Reply:
> Also you can't limit who can create/edit specific setups, i.e. if you allow me to create a Composition I can provision anything.

Each `Composition` is strongly coupled with one kind of composite resource (e.g. a `Wordpress` or a `MySQLInstance`), and doesn't do anything until it's invoked by a instance of that composite resource. In order to abuse a `Composition` you must have RBAC access to author a `Composition` _and_ the kind of composite resource that may invoke that composition.

I imagine no controller would actually reconcile a `Composition` directly. It would instead effectively specify a template to be used by a controller that knew it was reconciling either an application or an infrastructure resource - for example the `kind: Wordpress` controller or the `kind: MySQLInstance` controller.

Presumably an application controller will be allowed to compose _any_ resource as long as it's namespaced and created within the namespace of the composite. i.e. A persona who can author a `Wordpress` and a `Composition` may (indirectly) create any kind of resource within any namespace in which they have RBAC permissions to create `Wordpress` instances.

Presumably an infrastructure controller will only be allowed to compose infrastructure resources (e.g. either primitive or other composite managed resources). i.e. A persona who can author a `MySQLInstance` and a `Composition` may (indirectly) create any kind of infrastructure resource. More concerningly, a persona who can author a `MySQLInstanceBinding` and a `Composition` may indirectly create any kind of infrastructure resource, by authoring a `Composition` that targets `MySQLInstance`.

I'm personally not particularly concerned about someone who has RBAC access to create instances of a particular kind of infrastructure and to define a `Composite` being able to leverage this power to (indirectly) create _any_ kind of infrastructure. I imagine wanting to restrict infrastructure operators to managing some but not all kinds of infrastructure will be a relative edge case.

Typically application operators (and developers) exist under tighter RBAC restrictions, so I agree that allowing them to create any namespaced resource or any infrastructure resource is concerning. We could distinguish between `ApplicationComposition` and `InfrastructureComposition` for RBAC purposes, but doing so would only prevent application operators from creating arbitrary kinds of infrastructure resource.

One interesting approach here might to be to add an optional `serviceAccountName` to `InfrastructureDefinition` and `ApplicationDefinition` - e.g.:

```yaml
apiVersion: crossplane.io/v1alpha1
kind: ApplicationDefinition
metadata:
  name: wordpresses.example.org
spec:
  crdSpecTemplate:
    # ...
  serviceAccountName: wordpresses.example.org
```

We'll need to instantiate a controller for each `ApplicationDefinition` (or `InfrastructureDefinition`). If this service account name were specified the controller watching for `Wordpress` would use its bearer token to authenticate to the API server, and thus be constrained by the RBAC permissions of that controller. This would allow the author of the `ApplicationDefinition` that defines `Wordpress` to use plain-old Kubernetes RBAC to restrict the kinds of resources a `Wordpress` may compose; the controller would fail if a `Wordpress` tried to compose a `Deployment` but was not allowed to.

Reply:
> I agree that introducing these additional concepts makes the latter design safer. Each resource kind can be natively validated by the API server. The separation of functionality into several distinct resource kinds enables granular RBAC roles, and reduces the blast radius of updating some resources. However, having written this all out, I am not convinced that the increased safety of this model justifies the increased cognitive complexity.

I see your point but I'm on the other side of the tradeoff; I think results of different composition classes differ so much that they deserve separation on `kind` level (by the way it's 9 kinds, you don't need to define cluster-scoped kubernetescluster and except static prov. you don't interact with 3 intermediary ones).

> I'm personally not particularly concerned about someone who has RBAC access to create instances of a particular kind of infrastructure and to define a Composite being able to leverage this power to (indirectly) create any kind of infrastructure. I imagine wanting to restrict infrastructure operators to managing some but not all kinds of infrastructure will be a relative edge case.

I can see that one can make a similar case for `ClusterRole` and `ClusterRoleBinding`; if you have them you can do anything **in the cluster**. In our case, the consequences of being able to compose anything is overreaching since you will suddenly have the permissions to create any cloud resource with any of the credential sets in the cluster. This is why I'm leaning on the side of security when we do those trade-offs. (case for `ClusterRole` still exists but we'd introduce yet another one)

> We'll need to instantiate a controller for each ApplicationDefinition (or InfrastructureDefinition). If this service account name were specified the controller watching for Wordpress would use its bearer token to authenticate to the API server, and thus be constrained by the RBAC permissions of that controller. This would allow the author of the ApplicationDefinition that defines Wordpress to use plain-old Kubernetes RBAC to restrict the kinds of resources a Wordpress may compose; the controller would fail if a Wordpress tried to compose a Deployment but was not allowed to.

That's an interesting angle and it seems to me that'd work. However, keep in mind that a hand-crafted `ClusterRole`, `ClusterRoleBinding`, `ServiceAccount` and a pod per each `InfrastructureDefinition` could be exhaustive.

Reply:
> a pod per each InfrastructureDefinition

Point taken, though for clarification I wasn't thinking there would be a pod per `InfrastructureDefinition`. It should be possible to do all of this from one pod at the expense of giving that pod permissions to read arbitrary service accounts.